HPCC-30583 Fix parquet column reads

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

dcamper commented:
@jackdelv The fact that XPATH is now supported may be worth a documentation update.  I suspect that this issue -- where a Parquet file contains a field name that does not cleanly map to an ECL attribute name -- will come up quite often, like it does with XML and JSON files.  It would be worth calling out to the ECL programmer that using XPATH as usual is the correct workaround.

dcamper commented on plugins/parquet/README.md:
```diff
@@ -32,7 +32,14 @@ The Parquet Plugin offers the following main functions:
 
 The Read function allows ECL programmers to create an ECL dataset from both regular and partitioned Parquet files. It leverages the Apache Arrow interface for Parquet to efficiently stream data from ECL to the plugin, ensuring optimized data transfer.
 
+In order to read a parquet file it is necessary to define the record structure of the file you intend to read with the names of the fields as stored in the parquet file and the type that you wish to read them as. It is possible for a parquet file to have field names that contain characters that are incompatible with the ECL field name definition. For example, ECL field names are case insensitive  ...
```
I think you should capitalize 'Parquet'.

jackdelv replied:
Fixed

dcamper reviewed (changes requested):
Reviewing only the README.  Good verbiage describing XPATH, just need to tweak the name of "Parquet".

jackdelv commented:
@dcamper I have updated the documentation to make it clearer to the user how to read a parquet file. Updated the example code to show an XPATH example.

jakesmith commented on plugins/parquet/parquetembed.cpp:
```diff
@@ -1337,7 +1336,8 @@ void ParquetRowBuilder::nextField(const RtlFieldInfo *field)
     }
-    auto column = result_rows->find(field->name);
+    (*array_visitor) = std::make_shared<ParquetArrayVisitor>();
+    auto column = result_rows->find(field->xpath ? field->xpath : field->name);
```
not new/future/question:

for records with many fields, this linear search by field name per column into an arrow::Array could be quite expensive by the looks of it.
I assume their positions in the array are invariant between calls to nextField for a given ParquetRowBuilder instance?
And are probably also invariant from row to row?

Can you prepopulate a map before ParquetRowStream::nextRow() is called with a table from RtlFieldInfo -> to position in the arrow::Array, to avoid these find's ?

jackdelv replied:
I am not sure what you mean by a table from RtlFieldInfo? The result_rows object is a prepopulated map with the field name of the column and an array containing the values for that column. You are correct in that the positions are invariant from row to row and between calls to nextField.

jakesmith replied:
I misread. I thought the code was performing a find in arrow::Array per field name.
I see now that it's a find in 'result_rows' which is an unordered_map, whose rhs is a arrow::Array.

jakesmith reviewed (approved):
@jackdelv - looks good.

One question not directly related to the changes here.

ghalliday reviewed (approved):
@jackdelv looks good.  Please squash and I will merge.

jackdelv committed:
HPCC-30583 Fix parquet column reads
